### ISSUE TITLE
Introduce cop to warn against usage of `$!` and `$@`.

### ISSUE THREAD

**Issue by ioquatix** [rubocop/rubocop#10204]
https://bugs.ruby-lang.org/issues/18083#note-21

Introduce a cop to warn against all usage of `$!` and `$@`. There is no safe usage except for perhaps `expression rescue $!`.

Usage of `$!` can be rewritten:

```ruby

begin
rescue => error
  raise
ensure
  if error ...
end
```

But Matz also said he disagree with such an approach. But at least it's more safe than `if $!`.

**Comment by andyw8** [comment]
Care should be taken for how this interacts with `Style/SpecialGlobalVars`.

**Comment by ioquatix** [comment]
@duerst I agree but the reality is some systems it is unavoidable, for example committing or aborting a database traction, we need to guarantee at least one option has happened, so either its:

```ruby
def with_transaction
  transaction = db.begin_transaction
  
  begin
    yield transaction
  rescue Exception
    transaction.abort!
  ensure
    transaction.commit! unless transaction.open?
  end
end
```

Not sure how else you can express this.

**Comment by eregon** [comment]
FYI IO#close does not raise so you can call it whether it's already closed or not.
Caring `file` being `nil` is avoided by ensuring only what needs to be in the `begin` is:
```ruby
file = File.open(...)
begin
  yield file
ensure
  file.close
end
```

Same for the transaction example, it should be an explicit `begin`.
In general `ensure/rescue` without an explicit `begin` is a trap and incorrect.

**Comment by ydakuka** [comment]
The same story.

I have the following code. 

```ruby
class Middleware
  def log_memory
    logger.info("PID: #{$$}")
  end
end
```

If I run rubocop with autocorrect option, I will get the code:

```ruby
# frozen_string_literal: true

require 'English'
class Middleware
  def log_memory
    logger.info("PID: #{$PROCESS_ID}")
  end
end
```

**Comment by ioquatix** [comment]
Just for reference, we found a good case study on this: https://github.com/ruby/net-ftp/pull/24

**Comment by ioquatix** [comment]
https://github.com/rubocop/rubocop/blob/master/lib/rubocop/cop/style/global_vars.rb#L60-L62

The GlobalVars explicitly enables things like `$!` and `$ERROR_INFO` but I think we should remove these as they are unsafe.

https://github.com/rubocop/rubocop/pull/10211 was proposed as a specific cop for the issues discussed here but it was rejected in favour of modifying GlobalVars.

What's the best way forward?

**Comment by ioquatix** [comment]
I had a go implementing this here: https://github.com/socketry/rubocop-socketry/commit/025f7c3d78b1f33b73eaf3a39bc071e3d7747f0e

**Comment by andyw8** [comment]
I asked Claude to create a PR, based on @ioquatix's work: https://github.com/rubocop/rubocop/pull/14909
